google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1616,17 +1617,24 @@ public void run() {
 
     void close() {
       synchronized (lock) {
+        closed = true;
         scheduledFuture.cancel(false);
         if (!running) {
           decrementPendingClosures(1);
         }
```

Comment:
This needs to wrapped in an if-not-closed block to prevent it from being called multiple times.

```suggestion
        if (!closed) {
          closed = true;
          scheduledFuture.cancel(false);
          if (!running) {
            decrementPendingClosures(1);
          }
        }
```

The above `decrementPendingClosures(1)` call could be executed multiple times from the `handleCreateSessionsFailure` method. The problem that is associated with that manifests itself in the following case (but can also cause problems in production environments):
1. Set the `SPANNER_EMULATOR_HOST` to a valid value and start an emulator instance.
2. Execute the tests that try to use a non-existing database and then close the client.

What could happen in that case was:
1. A database client with a session pool is created.
2. The session pool tries to create sessions on the emulator. These are now in flight on the emulator (i.e. the client has not received any responses yet).
3. The client application closes the database client and session pool. As the session pool has not yet received any answers from the emulator, it assumes that 100 sessions are still being created (MinSessions=100) and that the maintainer is valid.
4. The `pendingClosure` count is therefore set to 101 during the `closeAsync` call.
5. The session creation fails and responds with a `DatabaseNotFoundException`. The `createSessions` method decreases the `pendingClosure` count with 25 for the sessions and then the `handleCreateSessionsFailure` decreases it with 1 for the pool maintainer.
6. Repeat the previous step three more times, and the value of `pendingClosure` is never exactly zero (but does reach a negative value).

Reply:
Thanks for the detailed explanation. I have updated the code accordingly.